### DISCUSSION TITLE
Customized ValidationErrorType

### DISCUSSION THREAD

**Discussion by dugenkui03** [graphql-java/graphql-java#2415] [General]
From [sepc](https://spec.graphql.org/draft/#sec-Language.Directives):
> Directives provide a way to describe alternate runtime execution and type validation behavior in a GraphQL document.

Directive provide a way to do some customized validation, but the `ValidationErrorType` used by `ValidationError` is Enum, and it can not be extends to describe customized validation error.

Is it a better way to make `ValidationErrorType` implements `ValidationErrorInterface`?
and we can extends validation error by implements `ValidationErrorInterface`.

**Comment by bbakerman** [comment]
Yes it is in a way - enums in Java alays lead to this problem of extension.

But actually the `ValidationErrorType` class itself is badly designed - why?  because extra attributes on a graphql error SHOULD be in the extensions not on the object itself


```
public class ValidationError implements GraphQLError {

    private final String message;
    private final List<SourceLocation> locations = new ArrayList<>();
    private final String description;
    private final ValidationErrorType validationErrorType;
    private final List<String> queryPath;
    private final Map<String, Object> extensions;
```

So really the extra validation indication should have been in an extensions object.  

Why is it like this - because this code was written in 2015 when the spec had little to say about error shape.

I suggest that IF you are writing custom validations that will produce errors, design your own error class and put any of the extra parameters into the extensions so that when they get serialised back to clients, it will be more spec compliant.


https://spec.graphql.org/June2018/#sec-Errors

> GraphQL services should not provide any additional entries to the error format since they could conflict with additional entries that may be added in future versions of this specification.

> Previous versions of this spec did not describe the extensions entry for error formatting. While non‐specified entries are not violations, they are still discouraged.

**Reply by dugenkui03** [reply]
Thank for you solution.
